readme.md
REDIPS.drag 5.3.1
============

## What is REDIPS.drag?

REDIPS.drag is a JavaScript drag and drop library focused on dragging table content and table rows. Here are minimal steps to enable content dragging or row dragging in HTML table:

* put **&lt;script type="text/javascript" src="redips-drag-min.js">&lt;/script>** to the head section
* initialize REDIPS.drag library: **&lt;body onload="REDIPS.drag.init()">**
* place table(s) inside **&lt;div id="redips-drag">** to enable content dragging
* place **&lt;div class="redips-drag">Hello World&lt;/div>** to the table cell

## Features

* dragging table content/row
* dropping modes: multiple, single, switch, switching, overwrite and shift
* 4 shift modes (horizontal1, horizontal2, vertical1 and vertical2)
* multiple drop options can drop to cell top or bottom
* using custom event handlers
* autoscroll
* content cloning
* support for scrollable DIV containers
* sticky tables and DIV containers
* unlimit nested tables support
* dropping rules (like allowed and forbidden table cells)
* animation (move element/row to the destination cell/row)
* "noautoscoll" option for drag containers
* "redips-nolayout" option for tables (table in DIV element can be dragged as any other content)
* takes care about rows with style display='none'
* support for dynamical creation of table layout (inner table inside DIV element can be moved and accept other DIV elements)
* works in touch devices (tested on Android 2.3.3)

## Public methods

* REDIPS.drag.init() - drag container initialization
* REDIPS.drag.initTables() - tables layout initialization
* REDIPS.drag.ajaxCall() - create AJAX call and handle response in callback function
* REDIPS.drag.clearTable() - deletes all DIV elements with "redips-drag" class name from table
* REDIPS.drag.cloneObject() - clones DIV element and returns reference of the cloned element
* REDIPS.drag.deleteObject() - method deletes DIV element (input parameter is DIV reference or element id)
* REDIPS.drag.emptyCell() - test if table cell contains DIV elements or remove elements from TD
* REDIPS.drag.enableDrag() - enables / disables one or all DIV elements in tables
* REDIPS.drag.enableTable() - enables / disables tables selected by class name
* REDIPS.drag.findCell() - returns data for first or last table cell (cell reference, row index and column index)
* REDIPS.drag.findParent() - returns reference of required parent element
* REDIPS.drag.getPosition() - returns DIV element position as array (tableIndex, rowIndex and cellIndex)
* REDIPS.drag.getScrollPosition() - returns scroll positions as array [ scrollX, scrollY ]
* REDIPS.drag.getStyle() - returns style value of requested object and style name
* REDIPS.drag.loadContent() - method loads content and inserts to the HTML table
* REDIPS.drag.moveObject() - method will calculate parameters and start animation (DIV element to the destination table cell)
* REDIPS.drag.relocate() - relocates all child nodes from source table cell to the target table cell
* REDIPS.drag.rowEmpty() - marks selected row as empty (tbl_id, row_idx, [color])
* REDIPS.drag.rowOpacity() - sets opacity to table row (el, opacity, color)
* REDIPS.drag.saveContent() - scans table and prepares table content as JSON object or query string
* REDIPS.drag.shiftCells() - method shifts table content horizontally or vertically (REDIPS.drag.shift_option defines the way of how content will be shifted)

## Documentation

A reference documentation with a list of public properties and methods contained in REDIPS.drag library.

* [REDIPS.drag documentation](http://www.redips.net/javascript/redips-drag-documentation/)
* [REDIPS.drag documentation - Appendix A](http://www.redips.net/javascript/redips-drag-documentation-appendix-a/)

## Demos

Live demo shows REDIPS.drag library in action: 

* [Drag and drop table content plus animation](http://www.redips.net/javascript/drag-and-drop-table-content-animation/)
* [JavaScript drag and drop plus content shift](http://www.redips.net/javascript/drag-and-drop-content-shift/)
* [Drag and Drop table rows](http://www.redips.net/javascript/drag-and-drop-table-row/)
* [Drag and drop table content](http://www.redips.net/javascript/drag-and-drop-table-content/)
* [jsFiddle examples](https://jsfiddle.net/user/dbunic/fiddles/)

## YouTube

Screencast of several examples contained in REDIPS.drag package:

* [Drag and drop table content with JavaScript](http://www.youtube.com/watch?v=8LtMk4DwLzA)
* [Drag and drop table rows with JavaScript](http://www.youtube.com/watch?v=5YuS6S0bFTU)
* [Drag and drop table content with JavaScript - School timetable](http://www.youtube.com/watch?v=ToJk1End4C4)



changes.txt
Changes for REDIPS.drag library

5.3.1 (2020-12-01)
- removeChild() method replaced with remove()
- with removing DOM elements now are unregistered event listeners as well

5.3.0 (2020-03-21)
- REDIPS.drag is checked with ESLint - cleaning and optimization (deleted some unused variables)
- added example29 "Move more DIV in column"

5.2.5 (2018-12-12)
- enabled loading HTML content to the DIV element in loadContent() method - thanks to Mark Landmann

5.2.4 (2017-04-16)
- added REDIPS.drag.only.divClass and REDIPS.drag.mark.exceptionClass rules
- now it's possible to allow defined DIV class name to "redips-only" or "redips-mark" TD
- fixed bug in clearTable() method

5.2.3 (2017-04-14)
- fixed dragging in case of folded TD
- folded TD can occure in case when row width is not big enough
- in that case row will be wrapped with cells in more than one line
- optimisations in handlerOnMouseDown() to determine index of clicked table, row and cell

5.2.2 (2017-04-01)
- added clearTable() - method deletes all DIV elements with "redips-drag" class name from table
- added example28 - Load table content (with AJAX)

5.2.1 (2017-03-08)
- modified saveContent() - method now sends class names and innerText of DIV elements
- added loadContent() - method loads table content from server side JSON output
- added error.loadContent(obj) - it's called in case if placing DIV element to the table is not possible (nonexistent table coordinates)
- added error.ajax(xhr)

5.2.0 (2017-02-06)
- added ajaxCall() method for making easier AJAX requests
- example03 changed to use ajaxCall() method
- example16 changed to use ajaxCall() method
- example22 changed to use ajaxCall() method
- example23 changed to use ajaxCall() method
- example24 changed to use ajaxCall() method
- code cleaning in examples 3, 16, 22, 23, 24

5.1.0 (2017-01-22)
- added jsFiddle examples to documentation

5.1.0 (2015-03-06)
- code cleaning - thanks to Naresh Jain
- drag class name and container id renamed to redips-drag
- mark class name renamed to redips-mark
- clone class name renamed to redips-clone
- only class name renamed to redips-only
- single class name renamed to redips-single
- trash class name renamed to redips-trash
- rowhandler class name renamed to redips-rowhandler
- row class name renamed to redips-row
- nolayout class name renamed to redips-nolayout
- noautoscroll class name renamed to redips-noautoscroll
- nodrag class name renamed to redips-nodrag
- in example03 added mysqli API
- fixed example11

5.0.9 (2014-08-29)
- fixed bug in findParent() method (problem with skip and search nodes)

5.0.8
- fixed bug (THEAD in table caused error when drag-n-drop table rows) - thanks to James Ostrowick

5.0.7
- code optimization (reduced number call of getScrollPosition() in boxOffset() method)

5.0.6 (2013-04-23)
- added example27 - TD content

5.0.6 (2013-03-07)
- small fixes

5.0.5 (2012-12-22)
- event.relocated renamed to event.relocateEnd
- added event.relocateBefore(div, to) - it's called before each element is relocated
- added event.relocateAfter(div, to) - it's called after each element is relocated
- added event.shiftOverflow - called after overflow happen in "shift" drop mode
- fixed emptyCell - if TD contains only text then it will be treated as empty cell
- animation.shift property renamed to shift.animation

5.0.4 (2012-12-20)
- modified enableDrag - method has option to accept node reference to enable / disable DIV elements inside that node
- fix inside relocate method (events were registered after DIV element was relocated and that should not be the case)
- added example26: jQuery dialog

5.0.3 (2012-12-19)
- modified emptyCell - now method has option to test target cell and to delete cell content

5.0.2 (2012-12-14)
- internal cleaning
- rowOption renamed to rowDropMode
- rowDropMode now has options: "before" (default), "after", "overwrite" and "switch"

5.0.1 (2012-12-06)
- "shift" method now takes care about marked cells - thanks to Mark Jacobs
- added shift.overflow property with the allowed values: "bunch", "delete" and "source" - thanks to Mark Jacobs 
- added example25: Math 4 kids

5.0.0 (2012-10-30)
- big code cleaning
- majority of properties, methods and event handlers are renamed (camelCase syntax is used)
- added option to set delete confirmation question (for deleting DIV element and row)
- initTables method exposed as public (needed for cases when table is dynamically added or removed)
- from now lib uses querySelectorAll for tables and div selection (here is compatibility table http://caniuse.com/queryselector)
- enableDrag uses CSS selector 
- added clone.sendBack option - if set to true, cloned DIV element could be returned to the source
- updated documentation
- added example21 Simple shift

4.7.6 (2012-10-22)
- added event handler myhandler_relocated() - event can be triggered after single call of relocate() method or after all DIV elements are shifted in "shift" mode

4.7.5 (2012-10-10)
- added "cloned" as optional input parameter to the myhandler_moved() event handler

4.7.4 (2012-09-18)
- added myhandler_final() event handler
- the color gets ignored in case a row is moved - bug fixed thanks to Rüdiger Kaatz
- added example24
- redips-drag.js renamed to redips-drag-source.js
- shift_after is converted to string parameter with options ("always", "delete" and "default")

4.7.3 (2012-09-14)
- fixed myhandler_dropped_before() in case of "overwrite" drag mode (bottom DIV element was deleted before calling myhandler_dropped_before event handler)  

4.7.2 (2012-08-21)
- added optional parameters target_row, source_table and source_row_index to the myhandler_row_dropped event handler

4.7.1 (2012-08-19)
- fixed bug with undefined target_cell in myhandler_switched() - thanks to Doc

4.7.0 (2012-08-17)
- added support for rowspan/colspan cells in "shift" mode

4.6.23 (2012-08-06)
- fixed bug in case of canceling drop in "switch" mode - thanks to Robin

4.6.22 (2012-07-25)

example03/index.php
<?php 
// include config with database definition
include('config_mysqli.php');
?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html>
	<head>
		<meta name="author" content="Darko Bunic"/>
		<meta name="description" content="Drag and drop table content with JavaScript"/>
		<meta name="viewport" content="width=device-width, user-scalable=no"/><!-- "position: fixed" fix for Android 2.2+ -->
		<link rel="stylesheet" href="style.css" type="text/css" media="screen"/>
		<script type="text/javascript">
			var redipsURL = '/javascript/drag-and-drop-example-3/';
		</script>
		<script type="text/javascript" src="../header.js"></script>
		<script type="text/javascript" src="../redips-drag-min.js"></script>
		<script type="text/javascript" src="script.js"></script>
		<title>Example 3: School timetable</title>
	</head>
	<body>
		<div id="main_container">
			<!-- tables inside this DIV could have draggable content -->
			<div id="redips-drag">
	
				<!-- left container -->
				<div id="left">
					<table id="table1">
						<colgroup>
							<col width="190"/>
						</colgroup>
						<tbody>
							<?php subjects() ?>
							<tr><td class="redips-trash" title="Trash">Trash</td></tr>
						</tbody>
					</table>
				</div><!-- left container -->
				
				<!-- right container -->
				<div id="right">
					<table id="table2">
						<colgroup>
							<col width="50"/>
							<col width="100"/>
							<col width="100"/>
							<col width="100"/>
							<col width="100"/>
							<col width="100"/>
						</colgroup>
						<tbody>
							<tr>
								<!-- if checkbox is checked, clone school subjects to the whole table row  -->
								<td class="redips-mark blank">
									<input id="week" type="checkbox" title="Apply school subjects to the week" checked/>
									<input id="report" type="checkbox" title="Show subject report"/>
								</td>
								<td class="redips-mark dark">Monday</td>
								<td class="redips-mark dark">Tuesday</td>
								<td class="redips-mark dark">Wednesday</td>
								<td class="redips-mark dark">Thursday</td>
								<td class="redips-mark dark">Friday</td>
							</tr>

							<?php timetable('08:00', 1) ?>
							<?php timetable('09:00', 2) ?>
							<?php timetable('10:00', 3) ?>
							<?php timetable('11:00', 4) ?>
							<?php timetable('12:00', 5) ?>
							<tr>
								<td class="redips-mark dark">13:00</td>
								<td class="redips-mark lunch" colspan="5">Lunch</td>
							</tr>
							<?php timetable('14:00', 7) ?>
							<?php timetable('15:00', 8) ?>
							<?php timetable('16:00', 9) ?>
						</tbody>
					</table>
				</div><!-- right container -->
			</div><!-- drag container -->
			<br/>
			<div id="message">Drag school subjects to the timetable (clone subjects with SHIFT key)</div>
			<div class="button_container">
				<input type="button" value="Save" class="button" onclick="redips.save()" title="Save timetable"/>
			</div>
		</div><!-- main container -->
	</body>
</html>